### PR TITLE
SOLR-15555 Improved caching on FilterQuery

### PR DESCRIPTION
**Pull request by madrob**
https://issues.apache.org/jira/browse/SOLR-15555

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/search/SolrCache.java
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/SolrCache.java:119]: Can also pass this along via an AtomicReference or possible via Lucene's Rethrows class if people would prefer to see that.

File: solr/core/src/test/org/apache/solr/search/TestRangeQuery.java
Comment (madrob) [inline solr/core/src/test/org/apache/solr/search/TestRangeQuery.java:402]: TODO: assert that createDocSet was only called once here instead of 4 times.

**Comment by madrob** [comment]
@sigram can't tag you as a reviewer, but would appreciate you taking a look as well, if you have time. Thanks!

File: solr/core/src/java/org/apache/solr/search/facet/UnInvertedField.java
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/facet/UnInvertedField.java:614]: This is inconsistent with the non-cache case.

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:812]: I like this `doCache` local var for clarity but otherwise I don't see why you are making any changes to this method.  It also appears getDocSetNC might get called twice if there's a cache miss?

File: solr/core/src/java/org/apache/solr/search/SolrCache.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/SolrCache.java:111]: Yes

File: solr/core/src/java/org/apache/solr/search/SolrCache.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/SolrCache.java:114]: I like this; it simplified some of the callers nicely.

File: solr/core/src/java/org/apache/solr/schema/RptWithGeometrySpatialField.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/schema/RptWithGeometrySpatialField.java:171]: Lets propagate it.  I am guessing maybe I was lazy here :-(

File: solr/core/src/test/org/apache/solr/request/TestUnInvertedFieldException.java
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/request/TestUnInvertedFieldException.java:99]: Nice

File: solr/core/src/java/org/apache/solr/search/facet/UnInvertedField.java
Comment (markrmiller) [inline solr/core/src/java/org/apache/solr/search/facet/UnInvertedField.java:624]: It was not likely trying to mimic ConcurrentHashMap#computeIfAbsent, as it is detrimental to the concurrency of the map due to it's locking. Lot's of other threads can be accessing the map, typically, also using computeIfAbsent. That is why the docs say to make value creation short and sweet. Uninverting fields doesn't really fit that definition. The old impl attempted to use a lock for every single key, the place holder, so that the thread(s) creating a field cache entry did not hamper unrelated map access the entire length of value creation. People have run into ridiculously bad contention with this with ConcurrentHashMap. The previous approach could be taken because the size of the map is essentially tiny given it's key/values are large data structures based on fields. ConcurrentHashMap was designed with requirements that it scale to very large sizes. Once Caffeine became the cache impl, it's computeIfAbsent impl has a mitigation that was added that is generally suitable for caching - do a get first and only computeIfAbsent if that returns null.

Comment (madrob) [reply solr/core/src/java/org/apache/solr/search/facet/UnInvertedField.java:624]: ConcurrentHashMap does locking on the Node entry, which is either the collision list (up to 8 keys) or a Fibonacci tree bin (when larger than 8 keys), so I think this risk of locking other consumers out is small, but not negligible. I've started moving toward a reservation node approach, which is similar in spirit, but hopefully much less complex in practice!

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java
Comment (madrob) [reply solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:812]: Very possible that I had other changes here before and then I had to back some of them out due to failing tests

File: solr/core/src/java/org/apache/solr/query/SolrRangeQuery.java
Comment (madrob) [reply solr/core/src/java/org/apache/solr/query/SolrRangeQuery.java:458]: A naked put in this context is a put without a previous call to get or without a computeIfAbsent. We typically don't want to just insert things into the cache, but in this case it's _probably_ ok because we would have already checked in SIS if the entry exists.

**Comment by madrob** [comment]
I was reading https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalAsyncCache.java#L88-L96 to check this, but somehow convinced myself that because we return the future then we're not blocking on the CHM compute call. But with your comment I realize that in the same thread case we would always be getting a _completed_ future. Need to think some more about whether the common fork-join pool is appropriate here or not.

**Comment by ben-manes** [comment]
The common pool is a nice default, but due to blocking I/O it is easy for the workers to be saturated and the system idle. You might want a dedicated pool to avoid surprising degradations due to foreign code's usage of this shared resource. Otherwise, if you wanted to avoid running the future on a thread pool and use a caller thread, you could complete the future manually using a similar pattern to your original code.

```java
var future = new CompletableFuture<V>();
var result = asyncCache.asMap().putIfAbsent(key, future);
if (result != null) {
  return result.join();
}
try {
  V value = // do work
  future.complete(value);
  return value;
} catch (Throwable t) {
  future.completeExceptionally(t);
  throw t;
}
```

File: solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java:842]: Sweet :-)

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (ben-manes) [inline solr/core/src/java/org/apache/solr/search/CaffeineCache.java:200]: Please upgrade to the latest Caffeine version. That oversight was fixed.

Comment (madrob) [reply solr/core/src/java/org/apache/solr/search/CaffeineCache.java]: It looks like there's an interesting discrepancy here still, where the get from compute will increment the cache lookups stat, while the get in put will not increment the stat. I'm reading the source for 2.9.4 - I don't think we're ready to go to 3.x yet, although I'll probably do that as soon as I'm done with this issue.

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (madrob) [inline solr/core/src/java/org/apache/solr/search/CaffeineCache.java:202]: @ben-manes this code path doesn't seem to record a hit in the cache stats for the "happy" case where the result is already completed normally. should I be using a different API?

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (ben-manes) [reply solr/core/src/java/org/apache/solr/search/CaffeineCache.java]: Stats gets very funky with a Map, e.g. what stats are recorded for `V replace(k, v)` and `boolean replace(k, old, new)`? In Guava we decided to not record stats for the `asMap()` view and not cherry-pick. That made sense since you could use `Cache.getIfPresent` and prefer a loading `get(k, loader)` over a `putIfAbsent`.

In Caffeine, I decided that `compute` was valuable enough to break that practice and record stats. I'd probably lean towards `asMap().get(key)` also recording because it's been a surprise that `getIfPresent` should be used instead, but we have Guava's history to justify keeping as is. Similarly putIfAbsent sounds good, but it gets too confusing about the "correct" stats. Thus the fallback is to tell users to not rely on ours and create their own to get what they deem to be the right metrics. Ours are therefore best effort, but fallible.

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (ben-manes) [reply solr/core/src/java/org/apache/solr/search/CaffeineCache.java:202]: I think stats on an async cache, especially with the `asMap()` view, is confusing. I don't know if I got it instrumented right. It tries to be good enough but either requires a deeper review to fix holes, or custom stats to match your desired reporting.

**Comment by madrob** [comment]
Rebased and squashed, would appreciate another look-over from folks that have participated. Working on additional performance testing before committing, but otherwise I think this is pretty close.

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (ben-manes) [inline solr/core/src/java/org/apache/solr/search/CaffeineCache.java:270]: remove redundant put?

**Comment by markrmiller** [comment]
I'm seeing some pretty bad test results in the latest work here - dozens to a 150 or so fails due to what looks initially like at least two separate NPE cases.

File: solr/benchmark/src/java/org/apache/solr/bench/search/FilterCache.java
Comment (markrmiller) [inline solr/benchmark/src/java/org/apache/solr/bench/search/FilterCache.java:47]: Overall, great benchmark. I do have some lessor issues with it, but I don't want to bog down this issue on it, I would not deal with anything in the benchmark here as it would hold up getting the issue into the release.

- On my desktop, the default run is fairly long. I have not finished the module doc, but I'd like to push default runs to be 10 minutes or less at most as you can easily crank things up or down as desired, but someone should be able to kick off a benchmark and get some kind of relatively reasonable feedback within a reasonable time and if they need to really engage with a benchmark then you alter settings on the command line. With a lot of benchmarks, if we don't manage this, running them all by default will take a very, very long time. This benchmark takes a bit over 20 min on my desktop.
- I don't think Threads.MAX is a great default. For one, it can often be simpler and less messy to start looking at multi threaded benchmarks with a handful of threads vs a ton. But more importantly, that takes into account hyper threaded cores - so on my desktop that uses 32 threads and at parts of the benchmark run, that can really clobber my machine for random small stretches. Of course that's not a bad idea, but you can configure a run like that easily via command line - it would be nice if the default run did not clobber my machine and just used a reasonably low number of threads.
- On my initial inspection, the results are not very consistent. I have some ideas on why that may be, but we want to have consistent results across runs.

File: solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformerFactory.java
Comment (markrmiller) [inline solr/core/src/java/org/apache/solr/response/transform/ChildDocTransformerFactory.java:185]: Not a fan of this myself.ll Impls change, new impls are added, it's declared to throw an IOException. Personally, I'd just turn it into a runtime io exception.

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (markrmiller) [inline solr/core/src/java/org/apache/solr/search/CaffeineCache.java:135]: You kind of narrowly dodge this being part of this issue, so more of an FYI, I'll second Ben's comments on this common thread pool usage.

It's a trappy and confusing use of common pool indicated to only impact a "cleanup" thread.

This is in fact a general executor for Caffeine, not simply the cleanup thread pool. In most cases, that is not a big deal, but the common pool is extremely easy to put into thread starvation, especially the common pool, if you perform any blocking tasks on it. There is an ugly workaround where you can indicate to the pool that you are going to block so it can compensate while you do so, but in general this pool is entirely meant for non blocking async type tasks. Where it get's trappy is that it's also used to do the work of loading for the async cache, which will often be doing IO, which is blocking. In this issue, async will use the current thread, so the case is avoid, but the code remains unclear and trappy and Caffeine can shift on where that executor is used in future versions given it's not a "cleanup" executor.

File: solr/core/src/test/org/apache/solr/search/TestSolrCachePerf.java
Comment (markrmiller) [inline solr/core/src/test/org/apache/solr/search/TestSolrCachePerf.java:132]: I would put it in one - there is no time out await below on that latch.

File: solr/solr-ref-guide/src/caches-warming.adoc
Comment (markrmiller) [inline solr/solr-ref-guide/src/caches-warming.adoc:82]: Nothing urgent here, but personally I'd like it to also be a little more clear that this issue can help solve the problem in the description of the JIRA. It's somewhat referenced with the mention of autowarming, but with no history of that issue, I'd have no idea that using this would help prevent that problem or why autowarming was someone another approach to the async option.

**Review by markrmiller — APPROVED** [review]
Great work @madrob. Good stuff.

I'm still looking at some low level differences from before and after this change, but nothing that critical or that should hold this up.

File: solr/core/src/java/org/apache/solr/search/CaffeineCache.java
Comment (madrob) [reply solr/core/src/java/org/apache/solr/search/CaffeineCache.java:135]: Yea, I detour around it, and this issue existed previously. We might want to investigate it in a follow up issue, but I'd rather not get into it now.

File: solr/solr-ref-guide/src/caches-warming.adoc
Comment (madrob) [reply solr/solr-ref-guide/src/caches-warming.adoc:82]: This already has

> The most noticable improvements will be seen when there are many queries racing to compute the same results before they are inserted into the cache.

Not sure what else you mean I should add?

**Comment by madrob** [comment]
@markrmiller I did some timing runs of this, and the setup cost per Trial works out to about 20s each. I have 2 benchmarks with 5 * 3 combinations each, so that is already 10 minutes, which is beyond your intuition's maximum, and we haven't even run any tests yet. Doing a smaller data set (1k docs instead of 30k) gets the setup to 13s each trial.

Maybe we only commit the freq=50 benchmark by default and let power users run with 2 and 98 by themselves? That would cut the default time in third.

**Comment by madrob** [comment]
Pushed the benchmark portion of this separately in c942c7242c2, rebased, will wait for precommit to finish one more time before merging.

**Comment by ben-manes** [comment]
nice! I think you said earlier you would upgrade to 3.x after this was merged. Please do that if possible.

**Pull request merged**
